### ISSUE TITLE
Max outgoing Ping count exceeded

### ISSUE THREAD

**Issue by tr8dr** [nats-io/java-nats-streaming#89]
I have used NATS, but am new to NATS streaming server.   I have written a publisher, that on startup, publishes perhaps 100,000 messages almost immediately (some starting state for a consumer).   At startup there may not be any clients listening to the streaming server, however it has been configured to keep a backlog of messages over a certain period.

To my surprise, I receive an exception "Max outgoing Ping count exceeded".   Are these pings to the NATS broker or is this looking for a response from one or more clients?
```
Exception in thread "GDAX Reconnect" java.lang.IllegalStateException: Max outgoing Ping count exceeded.
	at io.nats.client.impl.NatsConnection.sendPing(NatsConnection.java:899)
	at io.nats.client.impl.NatsConnection.flush(NatsConnection.java:850)
	at io.nats.streaming.StreamingConnectionImpl.publish(StreamingConnectionImpl.java:362)
	at io.nats.streaming.StreamingConnectionImpl.publish(StreamingConnectionImpl.java:294)
	at org.mmaker.apps.ticker.impl.NATSStreamingPublisher.publish(NATSStreamingPublisher.kt:67)
	at org.mmaker.apps.ticker.impl.NATSStreamingPublisher.publish(NATSStreamingPublisher.kt:86)
	at org.mmaker.apps.ticker.TickerPublisherKt.onUpdate(TickerPublisher.kt:47)
	at org.mmaker.apps.ticker.TickerPublisherKt$main$2.invoke(TickerPublisher.kt:105)
	at org.mmaker.apps.ticker.TickerPublisherKt$main$2.invoke(TickerPublisher.kt)
        ...
```

**Comment by sasbury** [comment]
Background-wise, the issue is that the streaming server flushes messages, to make sure they got to the server. The flush call uses a Ping msg to the server, which the server responds with Pong, telling the client that the flush succeeded. In general the client library only wants to have a small number of pings in flight to the server (ie wait for a limited number of pongs), because if that gets backed up it could indicate a problem.

Which version of the streaming library are you using?

**Comment by tr8dr** [comment]
I have been using the following snapshots, as the nats streaming was code incompatible with the latest NATS java release.

compile "io.nats:jnats:2.0.1-SNAPSHOT"
compile "io.nats:java-nats-streaming:2.0-SNAPSHOT"

Have not been able to get nats streaming to work correctly on the pubisher side.  (I have not had an issue with NATS itself, which I tried before moving to the streaming server).

**Comment by sasbury** [comment]
Hm...

So the publish call you are making should block until the flush returns, which should prevent lots of PINGs getting out there before a PONG comes back. The default max pings allowed is set to 2, which is pretty low, so if you were using threads maybe to publish the initial messages I could see how something could happen. Can you tell me a bit more about the publisher?

**Comment by tr8dr** [comment]
The publisher is using multiple threads.  I saw that the API does use a concurrent queue so thought this would be fine.   I could synchronize, avoiding multiple incoming overlapping messages, if you think this would help.

**Comment by sasbury** [comment]
The issue is more that the publishers are going faster than the flush's can get through. I am not sure of a good general solution in the library (although open to suggestions.) In the short term you could either:

* synchronize that initial herd of messages (in normal mode i think you will be ok, as the messages flow quickly.)
* adjust the number of max outgoing pings allowed by the connection. This is a setting in the nats connection's options, so you would have to manually create the nats connection for the streaming connection
* slow down the herd somehow (I don't like this one since it smells of sleep statements)

I had another idea, that may or may not help. You could use multiple connections at the start to pre-populate these streams. I don't like this idea much, but it is an idea. Each nats connection has that max pings limit.

I think synchronizing the initial herd is safest, but in the short term you could adjust the max pings to make sure your client works.

If you have any thoughts on how the library could help let me know.

**Comment by tr8dr** [comment]
I am not sure why the streaming server would care about bursty activity (which happens at the start of my connection and randomly thereafter).  I guess more fundamentally, what is the ping / pong used for.  Is this a 1:1 ack for each message sent?   For throughput wouldn't it make sense to:

1. send as many messages downstream as possible (without acks)
2. keep messages sent on client side until can guarantee that the sends have completed, but avoid requiring a ping/pong for each message
3. broker sends ack on sequence #s seen since last ack

Forgive me if I am grossly misrepresenting what may be occurring.   It did look like there is a ping/pong per message, but I may have misunderstood.

**Comment by sasbury** [comment]
right now it ping/pongs each message to insure it gets to the gnatsd, I will look at what other language clients do (i just adopted this one) and see if we can replicate something faster.

**Comment by tr8dr** [comment]
Can I make publisher synchronous or partially synchronous (as opposed to using futures / running out of pending pongs)?   Perhaps synchronous after being backlogged by K pongs?

**Comment by sasbury** [comment]
it is basically sync now, because of the flush. You can set the max outgoing pongs which would let more of them go before the max is exceeded (but you can't reset it after that, which is why I don't like that as a long term plan).

**Comment by sasbury** [comment]
talked to some of the other members of the team, i think i can improve on this and remove the ping/pong. will update when I get a fix in.

**Comment by tr8dr** [comment]
thanks

**Comment by sasbury** [comment]
Ok, that flush was excessive. I looked at it more and it was overkill from my translation to 2.0 of the client library.  I removed the flush call which should fix the max ping problem for you.

Of course I forgot to put Fixes in my commit, but should be fixed as of cfe953deb188743490b2106b55ac35e7ad8c9d2e. Travis is running now, and it should update the snapshot. Leaving this issue open until the final release, but if you use the snapshot tomorrow it should be good.

**Comment by tr8dr** [comment]
Thanks for the adjustment on the branch.  Perhaps off topic here, In terms of getting a stable NATS streaming server (production or near production ready) what version combination of NATS and NATS streaming server java should I be using?

**Comment by sasbury** [comment]
The 2.0.0 version of NATS is new, but is heavily tested and appears very fast and stable so far. As with any recent release it is hard to guarantee anything. The older release had a number of outstanding issues, so I would lean toward the new one. (I am planning a 2.0.1 this week with a couple fixes although none are stability related.)

For streaming, the 2.0.0 version (updated to match the new client API) is basically the same code, modulo changes required to use the new API. (And the mistaken flush call i just removed). The new version is less thread intensive, and can share an underlying io.nats.client.Connection to allow better thread control when multiple threads are needed. Again, the version is new (I plan to release this week.) So it is really your call, but I would use the newer libraries.

**Comment by sasbury** [comment]
Should be fixed in 2.0.1 (but i forgot to put in the fix comment)
